python/tvm/script/parser/core/parser.py

```diff
@@ -150,7 +150,7 @@ def add(self, var: str, value: Any, allow_shadowing: bool = False):
             The options of whether variable shadowing allwed for this variable.
         """
         # Skip if the key and value are equal to those in the var_table
-        if self.name2value[var] and self.name2value[var][-1] == value:
+        if self.name2value[var] and self.name2value[var][-1] is value:
             return
```

Comment:
Thanks for pointing out the issue! I believe either way might not be the most accurate, because it's possible that `self.name2valuep[var]` is a python integer or so, which cannot be compared using `is`. We might want to dispatch comparison according to different types

Reply:
Thanks for your suggestion. I have done some trials in the python terminal according to your concern. Let's have a look:
```
# python3
Python 3.8.13 (default, Apr 19 2022, 00:53:22) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 1 is 1
<stdin>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
True
>>> a = 1
>>> a is 1
<stdin>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
True
>>> a = 1
>>> b = 1
>>> a is b
True
>>> a = [1, 2, 3]
>>> b = 1
>>> a[0] is b
True
>>> b = [1, 2, 3]
>>> a [0] is b[1]
False
```
According to the above output, we can find that there will be a warning if we use literal directly, but here it is a variable/list/dict which in your concern contains a literal. It looks like that python differentiates literal from variables with `literal` value. In our case, it belongs to the latter. So maybe no problem here with "is".

There might be some other scenarios I didn't cover, feel free to point out.

Reply:
I believe the `is` operator checks reference equality rather than value equality. For integers, it will just check equality, so @lightzhan-intellif is correct. Whether it's preferred style is another question

Reply:
I do not have particular opinion on which style we should go for, but just wanted to point out the implication of the switch:

- `is` checks referential/pointer equality, or in python's term, identity, where it returns True only when `id(lhs) == id(rhs)`. It could depend on certain underlying implementation of the system, for example:

```python
>>> a = 257
>>> b = 257
>>> a is b
False
>>> a = 256
>>> b = 256
>>> a is b
True
```

- `==` checks equality that could be potentially overloaded, for TVM objects, it's using TVM's address comparison `.same_as()` rather than python's builtin `id()` method which is used in `is` operator. However

The implication of switching from `==` to `is` means that it bypasses TVM's `.same_as()` method, which at the moment I am not quite certain is suitable for broad usecases.

Therefore, how about we do the following: if `lhs` and `rhs` are numpy arrays, then we use numpy-specific behavior (e.g. elementwise equality), but otherwise we still use `==`.

Reply:
Yea, I have updated the code.